### PR TITLE
Gracefully shutdown VTGate instances

### PR DESCRIPTION
**Pull request by davidpiegza**
## Description

Currently, during a VTGate shutdown it's possible for established connections to execute new queries, those queries will fail with `vttablet: Connection Closed errors` (see also [this comment](https://github.com/vitessio/vitess/issues/12079#issuecomment-1753479869)).

This PR adds a check in MySQL server plugin if the connection is shutting down and returns a `ERServerShutdown` error.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/12079

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/conn.go [hunk 1]
```diff
@@ -1707,7 +1707,15 @@ func (c *Conn) IsMarkedForClose() bool {
...
+}
+
 // GetTestConn returns a conn for testing purpose only.
 func GetTestConn() *Conn {
 	return newConn(testConn{})
 }
+
+func GetTestServerConn(listener *Listener) *Conn {
+	return newServerConn(testConn{}, listener)
+}
```
Comment (dbussink) [inline go/mysql/conn.go:1721]: Can this function (and `GetTestConn` above it) be moved to `conn_flaky_test.go` instead? So it's in a proper `_test.go` file?

Comment (GuptaManan100) [reply go/mysql/conn.go:1721]: Yes functions in test files can only be used locally. They aren't exported.

From the Internet - 
```
Files matching *_test.go are only compiled when testing the package they are part of. If you are testing package A that uses package B, you won't have access to the _test.go code from package B.
```

File: go/mysql/conn.go [hunk 2]
```diff
@@ -1707,7 +1707,15 @@ func (c *Conn) IsMarkedForClose() bool {
...
+func (c *Conn) IsShuttingDown() bool {
+	return c.listener.shutdown.Load()
+}
+
 // GetTestConn returns a conn for testing purpose only.
 func GetTestConn() *Conn {
 	return newConn(testConn{})
 }
+
+func GetTestServerConn(listener *Listener) *Conn {
```
Comment (GuptaManan100) [inline go/mysql/conn.go:1719]: I know the function already says `Test` but maybe it is a good idea to add a comment something like - 
```
// GetTestServerConn is only meant to be used for testing. It creates a server connection using a testConn and the provided listener.
```
And I would move this function to the `conn_fake.go` file. That files defines the `testConn` struct too and it contains all the test code we keep in the package.

**Review by GuptaManan100 — APPROVED** [review]
Rest LGTM!

File: go/mysql/conn.go (hunk 2 above)
Comment (davidpiegza) [reply go/mysql/conn.go]: @GuptaManan100 Thanks! I moved this and the other `GetTestConn()` function to `conn_fake.go`.

**Review by dbussink — APPROVED** [review]

**Comment by davidpiegza** [comment]
Would it be possible to backport this to latest v18 as well (after release)?

**Review by frouioui — APPROVED** [review]

**Pull request merged**